src/c77_menuFOC.hpp

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#if (DISPLAY_TYPE > 0) && (FOCUS_STEPPER_TYPE != STEPPER_TYPE_NONE)
+
+// HIGHLIGHT states allow you to pick one of the sub functions.
+enum
+{
+  HIGHLIGHT_FOCUS_FIRST = 1,
+  HIGHLIGHT_FOCUS_ADJUSTMENT = 1,
+
+  HIGHLIGHT_FOCUS_LAST = HIGHLIGHT_FOCUS_ADJUSTMENT
+};
+
+#define FOCUS_ADJUSTMENT 90
+
+byte focState = HIGHLIGHT_FOCUS_FIRST;
+byte rateIndex = 3;
+
+bool processFocuserKeys()
+{
+  lcdButton_t key;
+  bool waitForRelease = false;
+  bool checkForKeyChange = true;
+
+  lcdButton_t currentButtonState = lcdButtons.currentState();
+
+  if (focState == FOCUS_ADJUSTMENT)
+  {
+    if (currentButtonState == btnUP)
+    {
+      if (!mount.isRunningFocus())
+      {
+        mount.focusContinuousMove(-1);
+      }
+    }
+    else if (currentButtonState == btnDOWN)
+    {
+      if (!mount.isRunningFocus())
+      {
+        mount.focusContinuousMove(1);
+      }
+    }
+  }
+
+  if (currentButtonState == btnNONE)
+  {
+    if (mount.isRunningFocus())
+    {
+      mount.focusStop();
+    }
+  }
+
+  if (checkForKeyChange && lcdButtons.keyChanged(&key))
+  {
+    waitForRelease = true;
+
+    switch (focState)
+    {
+    case HIGHLIGHT_FOCUS_ADJUSTMENT:
+      if (key == btnSELECT)
+      {
+        focState = FOCUS_ADJUSTMENT;
+      }
+      if (key == btnRIGHT)
+      {
+        lcdMenu.setNextActive();
+      }
+
+      break;
+
+    case FOCUS_ADJUSTMENT:
+    {
+      // UP and DOWN are handled above
+      if (key == btnSELECT)
+      {
+        focState = HIGHLIGHT_FOCUS_ADJUSTMENT;
+      }
+      else if (key == btnRIGHT)
+      {
+        rateIndex = adjustClamp(rateIndex, 1, 0, 3);
+        mount.focusSetSpeedByRate(rateIndex + 1);
+      }
+      else if (key == btnLEFT)
+      {
+        rateIndex = adjustClamp(rateIndex, -1, 0, 3);
+        mount.focusSetSpeedByRate(rateIndex + 1);
+      }
+    }
+    break;
+    }
+  }
+  return waitForRelease;
+}
+
+void printFocusSubmenu()
+{
+  char scratchBuffer[20];
+  if (focState == HIGHLIGHT_FOCUS_ADJUSTMENT)
+  {
+    lcdMenu.printMenu(">Focus Adjust");
+  }
+  else if (focState == FOCUS_ADJUSTMENT)
+  {
+    strcpy(scratchBuffer, "Rate:  1 2 3 4 *");
+    scratchBuffer[6 + rateIndex * 2] = '>';
```

Comment:
There's got to be a better way to do this... This looks so brittle. Maybe split up the `Rate:` string from the options or something?

Reply:
We're trying to minimize memory allocations like ones that would result from string concatenations (prevent heap fragmentation). This is the best way.

Reply:
Yeah I know memory fragmentation is a bitch on constrained systems like this. Maybe I'll figure out a clean & performant way we can do string manipulation within menu option selections.